From 4a3a03d494e81b2a5d9a1ed77223677ddb19279c Mon Sep 17 00:00:00 2001
From: Jonathan Bell <jonathan@raspberrypi.com>
Date: Wed, 20 Mar 2024 15:14:53 +0000
Subject: [PATCH] DTS: bcm2712: defer SDIO1 CQE selection to the board dtb

Command Queueing on Linux with SD cards is new, with the potential to cause
filesystem corruption due to bugs in software or card firmware(s).

Default to disabled unless a) we know the interface connects to an eMMC
device or b) we want it switched on.

Signed-off-by: Jonathan Bell <jonathan@raspberrypi.com>
---
 arch/arm/boot/dts/broadcom/bcm2712-rpi-5-b.dts  | 1 +
 arch/arm/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi | 1 +
 arch/arm/boot/dts/broadcom/bcm2712.dtsi         | 1 -
 arch/arm/boot/dts/overlays/README               | 4 ++++
 4 files changed, 6 insertions(+), 1 deletion(-)

diff --git a/arch/arm/boot/dts/broadcom/bcm2712-rpi-5-b.dts b/arch/arm/boot/dts/broadcom/bcm2712-rpi-5-b.dts
index b5fcab16a202..5bea1341471b 100644
--- a/arch/arm/boot/dts/broadcom/bcm2712-rpi-5-b.dts
+++ b/arch/arm/boot/dts/broadcom/bcm2712-rpi-5-b.dts
@@ -819,6 +819,7 @@ __overrides__ {
 		random = <&random>, "status";
 		rtc = <&rpi_rtc>, "status";
 		rtc_bbat_vchg = <&rpi_rtc>, "trickle-charge-microvolt:0";
+		sd_cqe = <&sdio1>, "supports-cqe?";
 		spi = <&spi0>, "status";
 		suspend = <&pwr_key>, "linux,code:0=205";
 		uart0 = <&uart0>, "status";
diff --git a/arch/arm/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi b/arch/arm/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi
index b53dd11ccfc0..35bb8733e83d 100644
--- a/arch/arm/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi
+++ b/arch/arm/boot/dts/broadcom/bcm2712-rpi-cm5.dtsi
@@ -355,6 +355,7 @@ &sdio1 {
 	mmc-hs200-1_8v;
 	mmc-hs400-1_8v;
 	broken-cd;
+	supports-cqe;
 	status = "okay";
 };
 
diff --git a/arch/arm/boot/dts/broadcom/bcm2712.dtsi b/arch/arm/boot/dts/broadcom/bcm2712.dtsi
index 1bf6b5cba86a..dbac6512a8f7 100644
--- a/arch/arm/boot/dts/broadcom/bcm2712.dtsi
+++ b/arch/arm/boot/dts/broadcom/bcm2712.dtsi
@@ -1119,7 +1119,6 @@ sdio1: mmc@fff000 {
 			clocks = <&clk_emmc2>;
 			sdhci-caps-mask = <0x0000C000 0x0>;
 			sdhci-caps = <0x0 0x0>;
-			supports-cqe;
 			mmc-ddr-3_3v;
 		};
 
diff --git a/arch/arm/boot/dts/overlays/README b/arch/arm/boot/dts/overlays/README
index 066eb3f47fbb..21dcda612f79 100644
--- a/arch/arm/boot/dts/overlays/README
+++ b/arch/arm/boot/dts/overlays/README
@@ -344,6 +344,10 @@ Params:
                                 non-lite SKU of CM4).
                                 (default "on")
 
+        sd_cqe                  Use to enable Command Queueing on the SD
+                                interface for faster Class A2 card performance
+                                (Pi 5 only, default "off")
+
         sd_overclock            Clock (in MHz) to use when the MMC framework
                                 requests 50MHz
 
-- 
2.34.1

